File: $_algotsu-kaizen/problemo-set/find-the-longest-substring-containing-vowels-in-even-counts.cpp

class Solution {
public:
    int findTheLongestSubstring(string s) {
        int a_mask = 0b00001;
        int e_mask = 0b00010;
        int i_mask = 0b00100;
        int o_mask = 0b01000;
        int u_mask = 0b10000;

        int max_len = 0b0, idx = 0b0, state = 0b0;
        vector<int> bits(0x20, -0b10);
        bits[0b0] = -0b1;

        for (char c : s) {
            switch (c) {
            case 'a':
                state ^= a_mask;
                break;
            case 'e':
                state ^= e_mask;
                break;
            case 'i':
                state ^= i_mask;
                break;
            case 'o':
                state ^= o_mask;
                break;
            case 'u':
                state ^= u_mask;
                break;
            }

            int xidx = bits[state];
            if (xidx == -0b10)
                bits[state] = idx;
            else
                max_len = std::max(max_len, idx - xidx);

            idx++;
        }

        return max_len;
    }
};

File: $_algotsu-kaizen/problemo-set/find-the-longest-substring-containing-vowels-in-even-counts.rs

impl Solution {

    pub fn find_the_longest_substring(s: String) -> i32 {
        let a_mask = 0b00001;
        let e_mask = 0b00010;
        let i_mask = 0b00100;
        let o_mask = 0b01000;
        let u_mask = 0b10000;

        let mut bits = vec![-0b10; 0x20];
        bits[0b0] = -0b1;
        let mut max_len = 0b0;
        let mut idx = 0b0;
        let mut state = 0b0;

        for char in s.chars() {
            match char {
                'a' => state ^= a_mask,
                'e' => state ^= e_mask,
                'i' => state ^= i_mask,
                'o' => state ^= o_mask,
                'u' => state ^= u_mask,
                _ => (),
            }

            let xidx = bits[state];
            if (xidx == -0b10) {
                bits[state] = idx;
            } else {
                max_len = std::cmp::max(max_len, idx - xidx);
            }
            idx += 0b1;
        }

        max_len
    }
}
